src/score_gen/score.rs

```diff
@@ -1,162 +1,542 @@
+    // with a set of constraints $\psi$ on  new variables $(u,y,z)$ such
+    // that $y=1$ if $\chi$ holds and $y=0$ otherwise.
+    // We introduce new variables $u,y,z$ that are computed as follows:
+    //
+    // u &= \chi(\mathbf{x});\\
+    // y &=\begin{cases}
+    // 0,& \text{if }u\neq 0;\\
+    // 1,& \text{if }u=0.
+    // \end{cases}\\
+    // z&=\begin{cases}
+    // 1/u,& \text{if }u\neq 0;\\
+    // 0,& \text{if }u=0.
+    // \end{cases}
+    let u = witness + b_prime - accumulator;
+    // Conditionally assign `1` or `0` to `y`.
+    let y = if u == Scalar::zero() {
+        composer.add_input(Scalar::one())
+    } else {
+        composer.add_input(Scalar::zero())
+    };
+    // Conditionally assign `1/u` or `0` to z
+    let mut z = composer.zero_var;
+    if u != Scalar::zero() {
+        // If u != zero -> `z = 1/u`
+        // Otherways, `u = 0` as it was defined avobe.
+        // Check inverse existance, otherways, err.
+        if u.invert().is_none().into() {
+            return Err(ScoreError::NonExistingInverse.into());
+        };
+        // Safe to unwrap here.
+        z = composer.add_input(u.invert().unwrap());
+    }
+    // We can safely unwrap `u` now.
+    // Now we need to check the following to ensure we can provide a boolean
+    // result representing wether the rangeproof holds or not:
+    // `u = Chi(x)`.
+    // `u * z = 1 - y`.
+    // `y * u = 0`.
+    let one = composer.add_input(Scalar::one());
+    composer.add_gate(
+        one,
+        chi_x_var,
+        composer.zero_var,
+        u,
+        -Scalar::one(),
+        Scalar::zero(),
+        Scalar::zero(),
+        Scalar::zero(),
+    );
+    let one_min_y = composer.add(
+        (Scalar::one(), one),
+        (-Scalar::one(), y),
+        Scalar::zero(),
+        Scalar::zero(),
+    );
+    let u_times_z = composer.mul(u, one, z, Scalar::zero(), Scalar::zero());
+    composer.assert_equal(one_min_y, u_times_z);
+    let y_times_u = composer.mul(u, one, y, Scalar::zero(), Scalar::zero());
+    composer.assert_equal(y_times_u, composer.zero_var);
+    // Constraint the result to be boolean
+    composer.bool_gate(y);
+    Ok(y)
 }
 
 // Given the y parameter, return the y' and it's inverse value.
-fn compute_y_primes(y: Scalar) -> (Scalar, Scalar) {
-    // Compute y'
-    let y_prime = y & (Scalar::from(2u64).pow(&[129, 0, 0, 0]) - Scalar::one());
-    // Compute 1/y' where `y' = y & 2^129 -1`
-    let inv_y_prime = y.invert().unwrap();
-    (y_prime, inv_y_prime)
+fn biguint_to_scalar(biguint: BigUint) -> Result<Scalar, Error> {
+    let mut bytes = [0u8; 32];
+    let biguint_bytes = biguint.to_bytes_le();
+    if biguint_bytes.len() > 32 {
+        return Err(ScoreError::InvalidScoreFieldsLen.into());
+    };
+    bytes[0..biguint_bytes.len()].copy_from_slice(&biguint_bytes[..]);
+    // Due to the previous conditions, we can unwrap here safely.
+    Ok(Scalar::from_bytes(&bytes).unwrap())
+}
```

Comment:
Also this one, shouldn't we have a `Scalar::from_slice` or something like that, that use the internal buffer of  the `Scalar` structure and take care of checking that the `len` is no greater than 32? It seems to be that this is an operation we (at least in the past) did more than few times.

Reply:
We can discuss that.
This check is done by the `from_bytes` function but returns a `CtOption` instead of an `Error`. That's why I did this function.